### PR TITLE
Add Monster Armor Types

### PR DESCRIPTION
**Pull request by txtsd**
## What does this do?
This adds monster armor types.

## How was it tested?
No tests. It was manually entered.

## Is there a Github issue this is resolving?
Yes #280 

## Did you update the docs in the API? Please link an associated PR if applicable.
No

## More
So right now it's a straight rip from the PDF. Let me know how you want the data actually formatted.
Here are the unique entries:
```
{'10 in humanoid form, 11 (natural armor) in bear and hybrid form',
 '10 in humanoid form, 11 (natural armor) in boar or hybrid form',
 '11 in humanoid form, 12 (natural armor) in wolf or hybrid form',
 '14 (natural armor), 11 while prone',
 '15 with _mage armor_',
 '16 with _barkskin_',
 'armor scraps',
 'barding scraps',
 'breastplate',
 'chain mail',
 'chain mail, shield',
 'chain shirt',
 'chain shirt, shield',
 'hide armor',
 'hide armor, shield',
 'leather armor',
 'leather armor, shield',
 'natural armor',
 'natural armor, shield',
 'patchwork armor',
 'plate',
 'scale mail',
 'scale mail, shield',
 'splint',
 'studded leather',
 'studded leather, shield'}
```
Words surrounded by underscores are in italics.

### CODE REVIEW

File: src/5e-SRD-Monsters.json [hunk 1]
```diff
@@ -6239,6 +6262,7 @@
     "subtype": null,
     "alignment": "unaligned",
     "armor_class": 14,
+    "armor_type": "14 (natural armor), 11 while prone",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:6265]: Things like this point out that our AC structure might be flawed?

File: src/5e-SRD-Monsters.json [hunk 2]
```diff
@@ -6418,6 +6442,7 @@
     "subtype": "any race",
     "alignment": "any alignment",
     "armor_class": 12,
+    "armor_type": "15 with _mage armor_",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:6445]: This isn't even an armor type?

File: src/5e-SRD-Monsters.json [hunk 3]
```diff
@@ -13507,6 +13573,7 @@
     "subtype": "any race",
     "alignment": "any alignment",
     "armor_class": 11,
+    "armor_type": "16 with _barkskin_",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:13576]: This isn't an armor type.

File: src/5e-SRD-Monsters.json [hunk 4]
```diff
@@ -13663,6 +13730,7 @@
     "subtype": null,
     "alignment": "neutral",
     "armor_class": 11,
+    "armor_type": "16 with _barkskin_",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:13733]: This isn't an armor type.

File: src/5e-SRD-Monsters.json [hunk 5]
```diff
@@ -24239,6 +24359,7 @@
     "subtype": "any race",
     "alignment": "any alignment",
     "armor_class": 12,
+    "armor_type": "15 with _mage armor_",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:24362]: This isn't an armor type.

File: src/5e-SRD-Monsters.json [hunk 6]
```diff
@@ -36716,6 +36909,7 @@
     "subtype": "human",
     "alignment": "neutral good",
     "armor_class": 10,
+    "armor_type": "10 in humanoid form, 11 (natural armor) in bear and hybrid form",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:36912]: This isn't an armor type? Though is a bit more confusing.

File: src/5e-SRD-Monsters.json [hunk 7]
```diff
@@ -36870,6 +37064,7 @@
     "subtype": "human",
     "alignment": "neutral evil",
     "armor_class": 10,
+    "armor_type": "10 in humanoid form, 11 (natural armor) in boar or hybrid form",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:37067]: Same with this.

File: src/5e-SRD-Monsters.json [hunk 8]
```diff
@@ -37226,6 +37421,7 @@
     "subtype": "human",
     "alignment": "chaotic evil",
     "armor_class": 11,
+    "armor_type": "11 in humanoid form, 12 (natural armor) in wolf or hybrid form",
```
Comment (bagelbits) [inline src/5e-SRD-Monsters.json:37424]: And this.

**Comment by bagelbits** [comment]
I think all of these:
```
'10 in humanoid form, 11 (natural armor) in bear and hybrid form',
 '10 in humanoid form, 11 (natural armor) in boar or hybrid form',
 '11 in humanoid form, 12 (natural armor) in wolf or hybrid form',
 '14 (natural armor), 11 while prone',
 '15 with _mage armor_',
 '16 with _barkskin_',
```
mean that how we do AC is flawed for monsters.

**Comment by txtsd** [comment]
Those are all how it is in the SRD, so suggestions as to how we should handle them are welcome!

For example: The Archmage entry says `Armor Class 12 (15 with mage armor)`

So should we drop these and add them later as a special case, or take care of them in this PR itself?

**Comment by txtsd** [comment]
Maybe we can do something like this:
```json
{
  "armor_class": 12,
  "armor_special": [
    {
      "type": "spell",
      "spell": "mage armor",
      "armor_class": 15
    }
  ]
}

{
  "armor_class": 14,
  "armor_type": "natural armor",
  "armor_special": [
    {
      "type": "condition",
      "condition": "prone",
      "armor_class": 11
    }
  ]
}

{
  "armor_class": 10,
  "armor_special": [
    {
      "type": "form",
      "form": "humanoid",
      "armor_class": 10
    },
    {
      "type": "form",
      "form": "bear",
      "armor_class": 11,
      "armor_type": "natural armor"
    },
    {
      "type": "form",
      "form": "hybrid",
      "armor_class": 11,
      "armor_type": "natural armor"
    },
  ]
}
```

**Comment by Redmega** [comment]
`armor_type` here seems to just imply "Whatever's in the parentheses after the armor class."

I think there's two paths this can take: Assume the above, and submit as is (so things like `"armor_type": "12 with mage armor"` will exist), or figure out an object structure.

My vote would be for something like this:
```json
{
"armor_type": null,
"armor_special": "15 with mage armor"
},
{
"armor_class": 14,
"armor_type": "natural armor",
"armor_special": "11 while prone"
}
```

fwiw I don't think we should have markdown italics in this field as none of the other fields have markdown (language, senses, etc)

File: src/5e-SRD-Monsters.json (hunk 6 above)
Comment (ogregoire) [reply src/5e-SRD-Monsters.json:36912]: The whole were* are confusing and barely possible to write in our structure.

**Comment by ogregoire** [comment]
We've already handled this kind of stuff in the past. The example with form is close to an acceptable solution.

I think that the armour class should be a structure instead of a flat number.

**Comment by bagelbits** [comment]
I was thinking probably something like this:
```json
{
  // Stuff
  "armor": {
    "armor_class": 12,
    "armor_special": [
      {
        "type": "spell",
        "spell": "mage armor",
        "armor_class": 15
      }
    ]
  }
  // Other stuff
}

{
  // Stuff
  "armor": {
    "armor_class": 14,
    "armor_type": "natural armor",
    "armor_special": [
      {
        "type": "condition",
        "condition": "prone",
        "armor_class": 11
      }
    ]
  }
  // Other stuff
}

{
  // Stuff
  "armor": {
    "armor_class": 10,
    "armor_special": [
      {
        "type": "form",
        "form": "humanoid",
        "armor_class": 10
      },
      {
        "type": "form",
        "form": "bear",
        "armor_class": 11,
        "armor_type": "natural armor"
      },
      {
        "type": "form",
        "form": "hybrid",
        "armor_class": 11,
        "armor_type": "natural armor"
      },
    ]
  }
  // Other stuff
}
```

**Comment by txtsd** [comment]
Okay looks like I'll be changing armor for every monster entry :sweat_smile:

**Comment by bagelbits** [comment]
Sorry about that. >___<

**Comment by bagelbits** [comment]
I'm trying to remember how we've handled stuff like this in the past.

**Comment by ogregoire** [comment]
The array of armors is the way to go. @bagelbits provided a correct draft example [here](https://github.com/bagelbits/5e-database/issues/280#issuecomment-707282371). I agree that notes may not be the best choice at the moment.

But we must find hints in how the armor class is defined. The AC is calculated based on different formulas. There is the "natural armor", but there are a lot of different AC formulas. Such as the formula for "light armor", the one for "medium armor", and so on. There are even two different AC formulas for "unarmored defense" for monk and barbarian.

So let's just be pragmatic and write that: formulas.

```
{
  "armor_class": [
    {
      "value": 12,
      "formula": "natural armor"
    },
    {
      "value": 15,
      "formulas": "mage_hand"
    }
  ]
}
```

```
{
  "armor_class": [
    {
      "value": 14,
      "formula": "natural armor"
    },
    {
      "armor_class": 11,
      "formula": "prone"
    }
  ]
}
```

**Comment by bagelbits** [comment]
For something like:
```
"formulas": "mage_armor"
```

Does it make sense for us to link the spell in some way?

**Comment by bagelbits** [comment]
An alternative approach building off of that could be:
```
"formula": "spell"
"spell": {
  "name": "Mage Armor",
  "index": "mage-armor",
  "url": "/api/spells/mage-armor"
}
```

**Comment by txtsd** [comment]
I don't think I'm going to have the time for this in the near future. Someone else can pick it up.

**Comment by Redmega** [comment]
Idk if I like `formula` for this, it implies we are including some type of math... I liked `armor_type` before, I feel like that's more in line with the way some other entries are labeled.

**Comment by ogregoire** [comment]
`formula` is not the name I chose, but the one WotC choose: https://dnd.wizards.com/articles/features/rules-answers-january-2016

You might prefer the word `method` which is also referenced, but there are definitely some maths involved...

**Comment by Redmega** [comment]
Yeah there's absolutely maths involved, but none that are evident from the data. We simply give the final number, and any contributing factors (i.e. spell, or worn armor, etc.) Hence my vote for just a `type` of some sort

**Comment by fergcb** [comment]
I'm also not a fan of `formula`. @Redmega is right that it implies that the data is supplying the instructions to calculate the AC, rather than the calculated AC or the condition that grants it. I would propose something like this, incorporating some of the previous suggestions:
```json
{
  "armor_class": [
    {
      "value": 12,
      "type": "natural"
    },
    {
      "value": 15,
      "type": "spell",
      "spell": {
        "name": "Mage Armor",
        "index": "mage-armor",
        "url": "/api/spells/mage-armor"
      },
      "desc": "15 with mage armor"
    }
  ]
}
```
```json
{
  "armor_class": [
    {
      "value": 14,
      "type": "natural"
    },
    {
      "value": 11,
      "type": "condition",
      "condition": "prone",
      "desc": "11 while prone"
    }
  ]
```

**Comment by bagelbits** [comment]
Now that lycanthropes are split apart, this should resolve a couple of the outliers.

**Comment by bagelbits** [comment]
I think I like @fergcb's approach. Though for conditions, I would use an APIRefernce shape to link to the Condition table. Also, for monsters with only one AC, we would essentially have to do this for consistency:
```json
{
  "armor_class": [
    {
      "value": 12,
      "type": "natural"
    }
  ]
}
```

As long as that's acceptable, I'm fine with it.

**Comment by fergcb** [comment]
@bagelbits More than acceptable! I'd forgotten about the Conditions collection, if I'm quite honest, but you know me and my fondness for my APIReferences :)

**Comment by bagelbits** [comment]
@fergcb Circling back to this finally, would the TS type for this look like:
```typescript
type ArmorClass = (ArmorClassNatural | ArmorClassArmor | ArmorClassSpell | ArmorClassCondition)[];
type ArmorClassNatural = {
  type: "natural";
  value: number;
  desc?: string;
}
type ArmorClassArmor = {
  type: "armor";
  value: number;
  armor_type: APIReference; // Equipment
  desc?: string;
}
type ArmorClassSpell = {
  type: "spell";
  value: number;
  spell: APIReference; // Spell
  desc?: string;
}
type ArmorClassCondition = {
  type: "condition";
  value: number;
  condition: APIReference; // Condition
  desc?: string;
}
```

**Comment by bagelbits** [comment]
Closing in favor of #518

**Pull request closed without merging**